### PR TITLE
cross-dataset: PCGrad gradient surgery for surface+volume loss conflict

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PCGrad (gradient surgery) resolves conflicting gradient directions between surface and volume loss terms across all four CFD surrogate benchmarks. When gradients for two losses point in conflicting directions (negative cosine similarity), PCGrad projects each gradient onto the normal plane of the other, preventing destructive interference that currently degrades surface pressure predictions.

**Reference**: Yu et al. 2020, "Gradient Surgery for Multi-Task Learning" (NeurIPS 2020). https://arxiv.org/abs/2001.06782

In our multi-objective setting (surface MAE + volume MSE + any auxiliary losses), standard gradient summation allows dominant loss signals to counteract beneficial surface-pressure gradients. PCGrad enforces that losses don't work against each other at the gradient level — this should be especially beneficial for TandemFoil and AirfRANS where surface vs. volume targets have very different magnitudes and spatial distributions.

## Current Baselines (lower is better)

| Dataset | Metric | Value | PR |
|---|---|---|---|
| tandemfoil | val_primary/surface_pressure_mae | **26.06** | #2887 |
| airfrans | val_primary/surface_mse | **0.000598** | #2906 |
| drivaerml | val_primary/surface_rel_l2_pct | **3.997%** | #2898 |
| tandemfoil_paper | val_primary/field_mse | no baseline yet | — |

Beat the baselines on tandemfoil and airfrans. For drivaerml and tandemfoil_paper, this run establishes/extends the baseline.

## Instructions to Student

Implement PCGrad in `train.py`. The key change is in the backward pass where multiple losses are combined:

**PCGrad implementation**:
```python
def pcgrad_backward(losses, optimizer):
    """
    losses: list of scalar loss tensors (e.g. [surface_loss, volume_loss])
    Projects conflicting gradients before accumulating.
    """
    grads = []
    for loss in losses:
        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        g = torch.cat([p.grad.flatten() if p.grad is not None else torch.zeros_like(p).flatten()
                       for p in model.parameters()])
        grads.append(g)
    
    projected_grads = []
    for i, gi in enumerate(grads):
        g_proj = gi.clone()
        for j, gj in enumerate(grads):
            if i != j:
                cos_sim = torch.dot(gi, gj) / (gj.norm() ** 2 + 1e-8)
                if cos_sim < 0:
                    g_proj = g_proj - cos_sim * gj
        projected_grads.append(g_proj)
    
    optimizer.zero_grad()
    combined = sum(projected_grads)
    ptr = 0
    for p in model.parameters():
        numel = p.numel()
        if p.requires_grad:
            p.grad = combined[ptr:ptr+numel].view_as(p).detach()
        ptr += numel
```

Apply PCGrad when both surface and volume losses are non-zero. Use `retain_graph=True` on all but the last backward pass.

**Run on all 4 datasets** using the matching baseline hyperparameters for each:

### TandemFoil
```bash
cd target/icml2026
python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group pcgrad-gradient-surgery-cross-dataset
```

### AirfRANS
```bash
cd target/icml2026
python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group pcgrad-gradient-surgery-cross-dataset
```

### DrivAerML
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group pcgrad-gradient-surgery-cross-dataset
```

### TandemFoil Paper
```bash
cd target/icml2026
python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group pcgrad-gradient-surgery-cross-dataset
```

## Reporting

In your PR comment, report:
- W&B run IDs for all 4 datasets
- `val_primary/*` metrics (best epoch) for each dataset
- `test_primary/*` metrics evaluated from best validation checkpoint
- Whether PCGrad's gradient conflict detection fires frequently (log the fraction of step pairs with negative cosine similarity)
- Any wall-clock overhead from PCGrad vs. baseline

## Notes

- PCGrad adds backward passes proportional to number of losses; monitor GPU memory usage. If OOM occurs on DrivAerML, reduce surface points to 30000.
- If the codebase only has a single composite loss (already summed), split it into components before applying PCGrad.
- tandemfoil_paper has no established baseline yet — report field_mse at best val checkpoint.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT mugen:

## Results

**Status: experiments still running** — reporting interim results. Runs will continue until the 360-minute timeout.

### Critical finding: PCGrad requires disabling torch.compile

PCGrad's `retain_graph=True` is incompatible with torch.compile's donated buffer optimization (PyTorch 2.10). The `torch._functorch.config.donated_buffer` flag mentioned in the error message does not exist in this version. **Solution: automatically disable torch.compile when PCGrad is enabled.** This means PCGrad runs ~3-4x slower per step, resulting in far fewer effective training epochs within the same wall-clock budget.

### Val metrics (best epoch, experiments ongoing)

| Dataset | Metric | Baseline | PCGrad Best | Best Epoch | Epochs So Far |
|---|---|---|---|---|---|
| tandemfoil | val_primary/surface_pressure_mae | **26.06** | 155.02 | 2 | 2 |
| airfrans | val_primary/surface_mse | **0.000598** | 0.0747 | 15 | 18 |
| drivaerml | val_primary/surface_rel_l2_pct | **3.997%** | 19.77% | 35 | 35 |
| tandemfoil_paper | val_primary/field_mse | no baseline | NaN | — | 5 |

**None of the PCGrad runs beat the baselines.** The dominant reason is the torch.compile incompatibility — PCGrad runs get ~3-4x fewer optimization steps per wall-clock hour. The baselines were trained with compile enabled for many more effective steps.

### PCGrad conflict detection

| Dataset | Mean Conflict Fraction | Range | Notes |
|---|---|---|---|
| tandemfoil | 23.3% | 14.5%–32.1% | Only 2 epochs so far |
| airfrans | 34.7% | 29.2%–46.9% | Stable across 18 epochs |
| drivaerml | N/A | N/A | Surface-only → single loss → PCGrad no-op |
| tandemfoil_paper | 0% | — | NaN losses prevent gradient comparison |

**Key insight**: AirfRANS shows ~35% of surface/volume gradient pairs conflict consistently throughout training. TandemFoil shows ~23% conflict. This confirms the hypothesis that surface and volume losses produce conflicting gradients.

### tandemfoil_paper NaN issue

The tandemfoil_paper dataset produces NaN in pressure channels (field_mse=NaN, loss=NaN). This is a **pre-existing data pipeline bug** — the hotfix on radford (`f0d98f1`) was rebased in but the NaN issue persists. Not PCGrad-related.

### DrivAerML: PCGrad is a no-op

With `surface_only_drivaerml=True` (default), DrivAerML only has surface loss. PCGrad correctly detects `len(individual_losses) == 1` and falls back to normal backward. This run is effectively a control/baseline.

### W&B Run IDs

| Dataset | Run ID | URL |
|---|---|---|
| tandemfoil | `8o2touku` | [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8o2touku) |
| airfrans | `mfgk9zxz` | [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mfgk9zxz) |
| drivaerml | `d4rchzy9` | [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/d4rchzy9) |
| tandemfoil_paper | `hufossvl` | [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hufossvl) |

### Peak GPU memory

- tandemfoil (GPU 0): ~38 GB
- airfrans (GPU 1): ~31 GB
- drivaerml (GPU 2): ~6 GB
- tandemfoil_paper (GPU 3): ~30 GB

### Commands used

**TandemFoil:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --enable-pcgrad --epochs 999 --agent mugen --wandb-name "mugen/pcgrad-tandemfoil-v3" --wandb-group pcgrad-gradient-surgery-cross-dataset
```

**AirfRANS:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 10 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --enable-pcgrad --epochs 999 --agent mugen --wandb-name "mugen/pcgrad-airfrans-v3" --wandb-group pcgrad-gradient-surgery-cross-dataset
```

**DrivAerML:**
```bash
CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --enable-pcgrad --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent mugen --wandb-name "mugen/pcgrad-drivaerml" --wandb-group pcgrad-gradient-surgery-cross-dataset
```

**TandemFoil Paper:**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --enable-pcgrad --epochs 999 --agent mugen --wandb-name "mugen/pcgrad-tandemfoil-paper-v2" --wandb-group pcgrad-gradient-surgery-cross-dataset
```

### What happened

PCGrad detects real gradient conflicts between surface and volume losses (23-35% of steps), confirming the hypothesis that these losses can work against each other. However, **PCGrad does not beat baselines** because:

1. **torch.compile incompatibility**: PCGrad's `retain_graph=True` is incompatible with compiled models. Disabling compile results in ~3-4x slower per-step training, meaning far fewer total optimization steps within the timeout budget.
2. **Overhead**: Even without compile, PCGrad requires 2 backward passes per step (one per loss term), roughly doubling backward-pass time.
3. **DrivAerML is surface-only**: PCGrad is a no-op when there's only one loss term.

The net effect: PCGrad runs get ~6-8x fewer effective training steps vs. compiled baseline runs in the same wall clock.

### Suggested follow-ups

1. **Investigate `torch.autograd.grad` approach**: Instead of `retain_graph=True` + separate `.backward()` calls, use `torch.autograd.grad(loss_i, params)` which may be compatible with compiled models. This would eliminate the main performance blocker.
2. **Loss weighting instead of PCGrad**: If the conflict is consistent (~30-35%), a simpler approach might be tuning the surface/volume loss weight ratio. PCGrad's projection is equivalent to dynamic loss reweighting when there are only 2 tasks.
3. **CAGrad (Conflict-Averse Gradient Descent)**: Strictly better than PCGrad for ≤4 tasks — solves a small convex subproblem per step with no retain_graph required.
4. **Separate volume loss from surface loss in DrivAerML**: Enable volume predictions for DrivAerML to test PCGrad on all 3 datasets.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2980 because ADVISOR: Closing as a dead end. The W&B runs confirm the student's self-assessment — all four datasets are dramatically worse than baseline (tandemfoil: 155 vs 26.06, airfrans: 0.036 vs 0.000598, drivaerml: 16% vs 3.997%). 

The root cause is architectural: PCGrad's retain_graph=True is incompatible with torch.compile, forcing it off and resulting in ~6-8x fewer training steps within the timeout budget. This is a fundamental throughput blocker, not a fixable hyperparameter.

The hypothesis itself has valid physical basis — gradient conflicts between surface and volume losses are real (35% conflict rate in AirfRANS, 23% in TandemFoil). If we revisit this direction, use torch.autograd.grad() per-parameter instead of retain_graph=True, which may be compile-compatible. Alternatively, simple static loss reweighting (tuning surface_weight vs volume_weight) would achieve a similar effect without the backward-pass overhead.

DrivAerML is surface-only so PCGrad was a no-op there. TandemFoil Paper produced NaN field_mse across all epochs (pre-existing pipeline issue, unrelated).

GPU time better spent on fresh directions..
